server/src/CompDBCache.cpp

```diff
@@ -0,0 +1,63 @@
+#include "CompDBCache.h"
+
+#include <sys/stat.h>
+
+#include <cassert>
+
+CompDBCache::CompDBCache()
+  : db_(nullptr), mtime_(0) {
+}
+
+CompDBCache::~CompDBCache() {
+  clear();
+}
+
+CXCompilationDatabase CompDBCache::fromDirectory(const std::string &buildDir,
+                                                 CXCompilationDatabase_Error *error) {
+  assert(error != nullptr);
+
+  const std::string jsonFilename = buildDir + "compile_commands.json";
+
+  int statRes = -2;
+
+#ifdef _WIN32
+  struct _stat st;
+  statRes = _stat(jsonFilename.c_str(), &st);
+#else
+  struct stat st;
+  statRes = stat(jsonFilename.c_str(), &st);
+#endif
+
+  if (statRes == 0) {
+    if (jsonFilename != filename_ || st.st_mtime != mtime_) {
+      clear();
+
+      db_ = clang_CompilationDatabase_fromDirectory(buildDir.c_str(), error);
+
+      if (*error == CXCompilationDatabase_NoError) {
+        filename_ = jsonFilename;
+        mtime_ = st.st_mtime;
+      }
+    } else {
+      // Using the cached compilation database.
+      // Set the provided error code to indicate success.
+      *error = CXCompilationDatabase_NoError;
+    }
+  } else {
+    // Not a JSON compilation database (or the file doesn't exist).
+    // Caching has not (yet) been implemented for other formats.
+    clear();
+    db_ = clang_CompilationDatabase_fromDirectory(buildDir.c_str(), error);
+  }
+
+  return db_;
```

Comment:
I believe we can have only one call to `clang_CompilationDatabase_fromDirectory()` if we refactor the code as follow (not tested):

```suggestion
  const std::time_t mtime = [&] {
#ifdef _WIN32
    struct _stat st;
    int statRes = _stat(jsonFilename.c_str(), &st);
#else
    struct stat st;
    int statRes = stat(jsonFilename.c_str(), &st);
#endif

    return statRes == 0 ? st.st_mtime : 0;
  }();

  if (jsonFilename == filename_ && st.st_mtime == mtime_) {
    // Using the cached compilation database.
    // Set the provided error code to indicate success.
    *error = CXCompilationDatabase_NoError;
    return db_;
  }

  clear();
  db_ = clang_CompilationDatabase_fromDirectory(buildDir.c_str(), error);

  if (mtime != 0 && *error == CXCompilationDatabase_NoError) {
    // looks like JSON compilation database, cache the results
    filename_ = jsonFilename;
    mtime_ = st.st_mtime;
  }

  return db_;
```

Reply:
I see where you're going. I've now re-organized so there's only one call to `clang_CompilationDatabase_fromDirectory()`.